gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -88,233 +142,123 @@ open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
     }
   }
 
-  override fun apply(target: Project) {
-    val extension = target.extensions.create("anvil", AnvilExtension::class.java)
-
-    // Create a configuration for collecting CodeGenerator dependencies.
-    val anvilConfiguration = target.configurations.maybeCreate("anvil").apply {
-      description = "This configuration is used for dependencies with Anvil CodeGenerator " +
-        "implementations."
-    }
+  override fun getCompilerPluginId(): String = "com.squareup.anvil.compiler"
 
-    val once = AtomicBoolean()
+  override fun getPluginArtifact(): SubpluginArtifact = SubpluginArtifact(
+    groupId = GROUP,
+    artifactId = "compiler",
+    version = VERSION
+  )
 
-    fun PluginManager.withPluginOnce(
-      id: String,
-      action: (AppliedPlugin) -> Unit
-    ) {
-      withPlugin(id) {
-        if (once.compareAndSet(false, true)) {
-          action(it)
-        }
-      }
-    }
+  private fun disablePreciseJavaTracking(variant: Variant) {
+    variant.compileTaskProvider.configure { compileTask ->
+      val result = CheckMixedSourceSet.preparePreciseJavaTrackingCheck(variant)
 
-    // Apply the Anvil plugin after the Kotlin plugin was applied. There could be a timing
-    // issue. Also make sure to apply it only once. A module could accidentally apply the JVM and
-    // Android Kotlin plugin.
-    target.pluginManager.withPluginOnce("org.jetbrains.kotlin.android") {
-      realApply(target, ANDROID, extension, anvilConfiguration)
-    }
-    target.pluginManager.withPluginOnce("org.jetbrains.kotlin.multiplatform") {
-      target.afterEvaluate {
-        // We need to run this in afterEvaluate, because otherwise the multiplatform extension
-        // isn't initialized.
-        val targets =
-          target.extensions.findByType(KotlinMultiplatformExtension::class.java)?.targets
-
-        when {
-          targets?.any { it.platformType == androidJvm } == true ->
-            realApply(target, MULTIPLATFORM_ANDROID, extension, anvilConfiguration)
-          targets?.any { it.platformType == jvm } == true ->
-            realApply(target, MULTIPLATFORM_JVM, extension, anvilConfiguration)
-          else -> throw GradleException(
-            "Multiplatform project $target is not setup for Android or JVM."
-          )
-        }
-      }
-    }
-    target.pluginManager.withPluginOnce("org.jetbrains.kotlin.jvm") {
```

Comment:
All these workarounds with applying the plugin once or checking which plugin is applied are no longer needed.